File: src/bun.js/test/expect/toBeOdd.zig
pub fn toBeOdd(this: *Expect, globalThis: *JSGlobalObject, callFrame: *CallFrame) bun.JSError!JSValue {
    defer this.postMatch(globalThis);

    const thisValue = callFrame.this();

    const value: JSValue = try this.getValue(globalThis, thisValue, "toBeOdd", "");

    this.incrementExpectCallCounter();

    const not = this.flags.not;
    var pass = false;

    if (value.isBigInt32()) {
        pass = value.toInt32() & 1 == 1;
    } else if (value.isBigInt()) {
        pass = value.toInt64() & 1 == 1;
    } else if (value.isInt32()) {
        const _value = value.toInt32();
        pass = @mod(_value, 2) == 1;
    } else if (value.isAnyInt()) {
        const _value = value.toInt64();
        pass = @mod(_value, 2) == 1;
    } else if (value.isNumber()) {
        const _value = JSValue.asNumber(value);
        if (@mod(_value, 1) == 0 and @mod(_value, 2) == 1) { // if the fraction is all zeros and odd
            pass = true;
        } else {
            pass = false;
        }
    } else {
        pass = false;
    }

    if (not) pass = !pass;
    if (pass) return .js_undefined;

    // handle failure
    var formatter = jsc.ConsoleObject.Formatter{ .globalThis = globalThis, .quote_strings = true };
    defer formatter.deinit();
    const value_fmt = value.toFmt(&formatter);
    if (not) {
        const received_line = "Received: <red>{f}<r>\n";
        const signature = comptime getSignature("toBeOdd", "", true);
        return this.throw(globalThis, signature, "\n\n" ++ received_line, .{value_fmt});
    }

    const received_line = "Received: <red>{f}<r>\n";
    const signature = comptime getSignature("toBeOdd", "", false);
    return this.throw(globalThis, signature, "\n\n" ++ received_line, .{value_fmt});
}

const bun = @import("bun");

const jsc = bun.jsc;
const CallFrame = bun.jsc.CallFrame;
const JSGlobalObject = bun.jsc.JSGlobalObject;
const JSValue = bun.jsc.JSValue;

const Expect = bun.jsc.Expect.Expect;
const getSignature = Expect.getSignature;

File: src/runtime/test_runner/expect/toBeOdd.rs
use bun_jsc::{CallFrame, JSGlobalObject, JSValue, JsResult};
#[allow(unused_imports)] use super::{Expect, JSValueTestExt};

impl Expect {
    #[bun_jsc::host_fn(method)]
    pub fn to_be_odd(&self, g: &JSGlobalObject, f: &CallFrame) -> JsResult<JSValue> {
        self.run_unary_predicate(g, f, "toBeOdd", |v| {
            if v.is_big_int32() {
                v.to_int32() & 1 == 1
            } else if v.is_big_int() {
                v.to_int64() & 1 == 1
            } else if v.is_int32() {
                v.to_int32().rem_euclid(2) == 1
            } else if v.is_any_int() {
                v.to_int64().rem_euclid(2) == 1
            } else if v.is_number() {
                let n = v.as_number();
                n.rem_euclid(1.0) == 0.0 && n.rem_euclid(2.0) == 1.0
            } else {
                false
            }
        })
    }
}
// ported from: src/test_runner/expect/toBeOdd.zig
